### PR TITLE
sig-algs: add support for RSA-PSS certs on server side

### LINKED ISSUE #512: Update test case test-sig-algs.py to work with RSA-PSS certificates
# Bug Report

## System Information

| Type                     | Version/Name |
| ------------------------ | ------------ |
| Operating System Name    | Fedora |
| Operating System Version | 29 |
| Python version           | python3-3.7.2-1.fc29.x86_64 |
| tlsfuzzer revision       | tlsfuzzer-0.0.0-alpha26 |

## Problem description

Test case  test-sig-algs.py does not work correctly when RSA-PSS certificates are used. Issues are most like caused by the fact that it expects a rsaEncryption cert on server side but there might be other issue too.

## Describe how to reproduce the problem

1. Create RSA-PSS certificates.
2. PYTHONPATH=tlsfuzzer python3 tlsfuzzer/scripts/test-sig-algs.py

## Expected behaviour

Test case should PASS (at least 'with RSA-PSS').

## Include errors or backtraces

  <details>
    <summary>Full test case output</summary><p>

```
sanity ...
OK

with RSA-PSS ...
Error encountered while processing node <tlsfuzzer.expect.ExpectServerHello object at 0x7f9ab705e0b8> (child: <tlsfuzzer.expect.ExpectCertificate object at 0x7f9ab705e0f0>) with last message being: <tlslite.messages.Message object at 0x7f9ab706af98>
Error while processing
Traceback (most recent call last):
  File "tlsfuzzer/scripts/test-sig-algs.py", line 477, in main
    runner.run()
  File "/tmp/tmp.ZEiRAYpxW0/tlsfuzzer/tlsfuzzer/runner.py", line 221, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Alert(fatal, handshake_failure)

rsa_pss_rsae_sha256 only ...
Error encountered while processing node <tlsfuzzer.expect.ExpectServerHello object at 0x7f9ab705e630> (child: <tlsfuzzer.expect.ExpectCertificate object at 0x7f9ab705e668>) with last message being: <tlslite.messages.Message object at 0x7f9ab706aac8>
Error while processing
Traceback (most recent call last):
  File "tlsfuzzer/scripts/test-sig-algs.py", line 477, in main
    runner.run()
  File "/tmp/tmp.ZEiRAYpxW0/tlsfuzzer/tlsfuzzer/runner.py", line 221, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Alert(fatal, handshake_failure)

rsa_pss_rsae_sha384 only ...
Error encountered while processing node <tlsfuzzer.expect.ExpectServerHello object at 0x7f9ab705eb70> (child: <tlsfuzzer.expect.ExpectCertificate object at 0x7f9ab705eba8>) with last message being: <tlslite.messages.Message object at 0x7f9ab706af28>
Error while processing
Traceback (most recent call last):
  File "tlsfuzzer/scripts/test-sig-algs.py", line 477, in main
    runner.run()
  File "/tmp/tmp.ZEiRAYpxW0/tlsfuzzer/tlsfuzzer/runner.py", line 221, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Alert(fatal, handshake_failure)

rsa_pss_rsae_sha512 only ...
Error encountered while processing node <tlsfuzzer.expect.ExpectServerHello object at 0x7f9ab70620f0> (child: <tlsfuzzer.expect.ExpectCertificate object at 0x7f9ab7062128>) with last message being: <tlslite.messages.Message object at 0x7f9ab75165f8>
Error while processing
Traceback (most recent call last):
  File "tlsfuzzer/scripts/test-sig-algs.py", line 477, in main
    runner.run()
  File "/tmp/tmp.ZEiRAYpxW0/tlsfuzzer/tlsfuzzer/runner.py", line 221, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Alert(fatal, handshake_failure)

MD5 first ...
Error encountered while processing node <tlsfuzzer.expect.ExpectServerHello object at 0x7f9ab70625f8> (child: <tlsfuzzer.expect.ExpectCertificate object at 0x7f9ab7062630>) with last message being: <tlslite.messages.Message object at 0x7f9ab706a710>
Error while processing
Traceback (most recent call last):
  File "tlsfuzzer/scripts/test-sig-algs.py", line 477, in main
    runner.run()
  File "/tmp/tmp.ZEiRAYpxW0/tlsfuzzer/tlsfuzzer/runner.py", line 221, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Alert(fatal, handshake_failure)

MD5 first, no SHA-1 ...
Error encountered while processing node <tlsfuzzer.expect.ExpectServerHello object at 0x7f9ab7062b38> (child: <tlsfuzzer.expect.ExpectCertificate object at 0x7f9ab7062b70>) with last message being: <tlslite.messages.Message object at 0x7f9ab706aba8>
Error while processing
Traceback (most recent call last):
  File "tlsfuzzer/scripts/test-sig-algs.py", line 477, in main
    runner.run()
  File "/tmp/tmp.ZEiRAYpxW0/tlsfuzzer/tlsfuzzer/runner.py", line 221, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Alert(fatal, handshake_failure)

no SHA-1 ...
Error encountered while processing node <tlsfuzzer.expect.ExpectServerHello object at 0x7f9ab70660b8> (child: <tlsfuzzer.expect.ExpectCertificate object at 0x7f9ab70660f0>) with last message being: <tlslite.messages.Message object at 0x7f9ab706afd0>
Error while processing
Traceback (most recent call last):
  File "tlsfuzzer/scripts/test-sig-algs.py", line 477, in main
    runner.run()
  File "/tmp/tmp.ZEiRAYpxW0/tlsfuzzer/tlsfuzzer/runner.py", line 221, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Alert(fatal, handshake_failure)

extra sigalgs ...
Error encountered while processing node <tlsfuzzer.expect.ExpectServerHello object at 0x7f9ab70665f8> (child: <tlsfuzzer.expect.ExpectCertificate object at 0x7f9ab7066630>) with last message being: <tlslite.messages.Message object at 0x7f9ab706aef0>
Error while processing
Traceback (most recent call last):
  File "tlsfuzzer/scripts/test-sig-algs.py", line 477, in main
    runner.run()
  File "/tmp/tmp.ZEiRAYpxW0/tlsfuzzer/tlsfuzzer/runner.py", line 221, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Alert(fatal, handshake_failure)

only undefined sigalgs ...
OK

empty sigalgs ...
Error encountered while processing node <tlsfuzzer.expect.ExpectAlert object at 0x7f9ab7066cf8> (child: <tlsfuzzer.expect.ExpectClose object at 0x7f9ab7066d30>) with last message being: <tlslite.messages.Message object at 0x7f9ab706aac8>
Error while processing
Traceback (most recent call last):
  File "tlsfuzzer/scripts/test-sig-algs.py", line 477, in main
    runner.run()
  File "/tmp/tmp.ZEiRAYpxW0/tlsfuzzer/tlsfuzzer/runner.py", line 223, in run
    node.process(self.state, msg)
  File "/tmp/tmp.ZEiRAYpxW0/tlsfuzzer/tlsfuzzer/expect.py", line 1412, in process
    raise AssertionError(problem_desc)
AssertionError: Expected alert description "decode_error" does not match received "handshake_failure"

fuzz length of sigalgs ...
OK

truncate sigalgs extension ...
OK

odd length of sigalgs ...
OK

padded sigalgs ...
OK

sanity ...
OK

version: 2

Test end
successful: 7
failed: 9
  'empty sigalgs'
  'extra sigalgs'
  'MD5 first'
  'MD5 first, no SHA-1'
  'no SHA-1'
  'rsa_pss_rsae_sha256 only'
  'rsa_pss_rsae_sha384 only'
  'rsa_pss_rsae_sha512 only'
  'with RSA-PSS'
```
   </p></details>

### PR DESCRIPTION
**Pull request by tomato42**
### Description

Update the `test-sig-algs.py` script so that it can be run against server with RSA-PSS key

### Motivation and Context

more complete coverage for rsa-pss support

fixes #512 

### Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md) document and my PR follows [change requirements](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md#change-requirements) therein
- [x] the changes are also reflected in documentation and code comments
- [x] all new and existing tests pass (see Travis CI results)
- [x] [test script checklist](https://github.com/tomato42/tlsfuzzer/wiki/Test-script-checklist) was followed for new scripts
- [x] new test script added to `tlslite-ng.json` and `tlslite-ng-random-subset.json`
- [x] new and modified scripts were ran against popular TLS implementations:
  - [x] OpenSSL
  - [x] NSS
  - [x] GnuTLS
- [x] required version of tlslite-ng updated in requirements.txt and README.md

---
This change is [Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/516)

### CODE REVIEW

File: scripts/test-sig-algs.py [hunk 1]
```diff
@@ -487,6 +460,16 @@ def main():
...
+    print("Check if server correctly selects signature algorithm for SKE")
+    print("Test to verify that Server Key Exchange is signed with safe")
+    print("and correct algorithms.")
+    print("Note that test expects server with support for both rsa_pss_rsae_*")
+    print("and rsa_pss_pss_* signatures, in other words, one with both")
+    print("rsaEncryption key in one certificate and rsasse-pss in second")
+    print("certificate. If there's only one certificate installed in server,")
+    print("some of the tests that advertise just one algorithm may need to be")
+    print("disabled.")
+
```
Comment (The-Mule) [inline scripts/test-sig-algs.py]: So when testing NSS for instance. Do we need two server certificates in NSS DB? Or should we execute rsa_pss_pss_* and  rsa_pss_rsae_* probes separately in two executions? Perhaps I am missing something truly obvious here but I must admit that I am a bit confused here.

Comment (tomato42) [reply scripts/test-sig-algs.py]: either option will work

the note is here as tlslite-ng doesn't support multiple certificates, but other implementations do

but even if the implementation does support multiple certificates set up at the same time, doing two runs is a valid test (at the same time, for full coverage, it does mean the test needs to be run three times)

Comment (The-Mule) [reply scripts/test-sig-algs.py]: Thanks. I still sense some fragility in the tests. I tried approach you suggested and did three executions:

1. Server uses NSS DB "for" rsa_pss_pss certificates - all tests but rsa_pss_rsae_* passed.
2. Server uses NSS DB "for" rsa_pss_rsae certificates - all but the following tests passed:  'empty sigalgs', 'extra sigalgs', 'MD5 first', 'MD5 first, no SHA-1', 'no SHA-1' and 'rsa_pss_pss_* only' ".
3. Server uses NSS DB with both certs - all test but the following passed: 'empty sigalgs', 'extra sigalgs', 'MD5 first', 'MD5 first, no SHA-1', 'no SHA-1'.

Comment (The-Mule) [reply scripts/test-sig-algs.py]: All issues resolved. Thanks!

**Comment by tomato42** [comment]
thanks for review!

**Pull request merged**